### PR TITLE
EMTF: Improve protection against corrupted LCT data

### PR DESCRIPTION
**Pull request by jiafulow**
#### PR description:

Due to the [incident](https://hypernews.cern.ch/HyperNews/CMS/get/tier0-Ops/2170.html) from the recent MWGR, we found out that the EMTF emulator might crash if it receives CSC LCTs with unexpected parameters. This PR aims to **improve protection against such corrupted LCT data**.

The specific cause of the crash was the corrupted LCT pattern number. This pattern number is used to address a LUT to retrieve the phi coordinate correction value. The LUT expects a pattern number between 0 and 10. However, due to data corruption, the pattern number might exceed 10, and this line: `ph_patt_corr_.at(pattern);` would throw.

Changes made:
- To prevent this from happening in the future, all the functions involving a LUT in `L1Trigger/L1TMuonEndCap/src/SectorProcessorLUT.cc` now use "try-catch" blocks to catch errors due to `at()`.
- In `L1Trigger/L1TMuonEndCap/src/PrimitiveSelection.cc`, the LCT parameters such as strip, wire, valid, pattern, and quality, are now checked against allowed values. If any of the LCT parameters are found to be corrupted, the LCT is rejected (i.e. it is not used in any subsequent steps in the track finder, including coordinate conversion). The warning messages have been expanded.
- The current logic that tries to patch the pattern number and the quality number has been removed.

#### PR validation:

Validated with:

- `runTheMatrix.py -l limited -i all -j10`. The only errors were coming from `DAS_ERROR`.
- Re-processing the raw data from the Run 336345 during MWGR, the EMTF emulator doesn't crash anymore.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

N/A

---
Notifications: @eyigitba

### CODE REVIEW

**Comment by srimanob** [comment]
@jiafulow 
I assume you need this also for 11_1, in case of next MWGR.

**Comment by boudoul** [comment]
> @jiafulow
> I assume you need this also for 11_1, in case of next MWGR.

Yes I requested a backport on Tier0 from the beginning- It is likely needed before the next MWGR in order to reconstruct the data that couldn't be reconstructed last week.. Let me recall here what I was saying on the Tier0 HN : ' Please make a PR in the master branch and also a backport in 11_1_X. At this point MWGR is over, so we will consider calmy the possibility to reconstruct data with your fix. '
 Let's discuss at the run meeting and then if needed I can make a request to PdmV during the PPD coordination .

**Comment by jiafulow** [comment]
@srimanob @boudoul I'll make a 11_1_X port immediately. I haven't done it because I thought that I should wait until this PR is accepted before doing a backport.

**Comment by jiafulow** [comment]
@srimanob @boudoul A backport to 11_1_X has just been submitted: #30841 .

**Comment by silviodonato** [comment]
please test

File: L1Trigger/L1TMuonEndCap/src/SectorProcessorLUT.cc [hunk 1]
```diff
@@ -138,23 +138,55 @@ void SectorProcessorLUT::read(bool pc_lut_data, int pc_lut_version) {
 }
 
 uint32_t SectorProcessorLUT::get_ph_init(int fw_endcap, int fw_sector, int pc_lut_id) const {
-  size_t index = (fw_endcap * 6 + fw_sector) * 61 + pc_lut_id;
-  return ph_init_neighbor_.at(index);
+  const uint32_t index = (fw_endcap * 6 + fw_sector) * 61 + pc_lut_id;
+  uint32_t entry = 0;
+
+  try {
```
Comment (davidlange6) [inline L1Trigger/L1TMuonEndCap/src/SectorProcessorLUT.cc:144]: please avoid try-catch in CMSSW.

Comment (jiafulow) [reply L1Trigger/L1TMuonEndCap/src/SectorProcessorLUT.cc:144]: Please suggest the alternative?

Comment (davidlange6) [reply L1Trigger/L1TMuonEndCap/src/SectorProcessorLUT.cc:144]: if ( index < ph_init_neighbor_.size() )?

Comment (jiafulow) [reply L1Trigger/L1TMuonEndCap/src/SectorProcessorLUT.cc:144]: Ok

**Comment by boudoul** [comment]
Hi @rekovic, please test and eventually approve this PR,  we need it to reconstruct data from RAW affected by the EMTF firmware issue during the last MWGR (and please do not forget to  also take care of the backport   #30841 , it is actually with 11_1 that data should be reconstructed  ). Thank you

**Comment by boudoul** [comment]
Ping @rekovic

**Comment by boudoul** [comment]
It is a week now without a movement from @rekovic ,  @cms-sw/l1-l2 please take actions. 
CC  @BenjaminRS , @davignon ,  @silviodonato, @dpiparo

**Comment by silviodonato** [comment]
please test

**Comment by silviodonato** [comment]
urgent

**Comment by jiafulow** [comment]
I checked the comparison results [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-07-26-2300+8b6b7b/37976/dqm-histo-comparison-summary.html). I don't think the 5 "failures" are related to this PR. One came from WF 1001.0, and four came from WF 10224.0.

- WF 1001.0: AlCaReco/EventInfo/processTimeStamp. I'm not sure why a timestamp is kept in a DQM histogram? Obviously the timestamp is going to be different.

- WF 10224.0: MessageLogger/Errors/modulesErrorsFound, MessageLogger/Errors/modules_errors, MessageLogger/Warnings/modulesWarningsFound, MessageLogger/Warnings/modules_warnings. Apparently, the numbers of errors and warnings for this particular job are different. I'm not sure if the logs are kept? But since the DQM plots are ok, I'd think these are not real errors/warnings.

**Comment by rekovic** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**